scalafix/input/src/main/scala/fix/FutureSrc.scala

```diff
@@ -0,0 +1,22 @@
+/*
+rule = "scala:fix.CrossCompat"
+ */
+package fix
+
+import scala.concurrent.{Future, ExecutionContext}
+import java.lang.Throwable
+
+class FutureSrc(fs: Future[Int])(implicit ec: ExecutionContext){
+  class E1 extends Throwable
+  class E2 extends Throwable
+
+  fs.onFailure {
+    case _: E1 => println("E1")
+    case _: E2 => println("E2")
```

Comment:
Can you also try with a different pattern? For instance a constructor pattern.

Reply:
I gave it a shot with function application (`fs.onSuccess(pf)`), but it's not possible since the rewrite would be:

```scala
fs.onComplete {
  case scala.util.Success(v) => pf.applyOrElse(v, Predef.identity[TYPE_OF_FS])
  case _ => ()
}
```

we don't know the type: `TYPE_OF_FS` since it could come from an expression.

Do you have other patterns in mind? Can you give me a code example for the `constructor pattern` ?

Reply:
~~~ scala
fs.onFailure {
  case _: E1 => println("E1")
  case E2(_) => println("E2")
  case E3 => println("E3")
}
~~~

Reply:
sure, I can add a test.